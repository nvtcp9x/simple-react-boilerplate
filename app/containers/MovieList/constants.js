export const URL_FILMS = 'http://localhost:3000/films'
export const FETCH_FILMS = 'FETCH_FILMS'
export const FETCH_FILMS_SUCCESS = 'FETCH_FILMS_SUCCESS'
export const FETCH_FILMS_FAIL = 'FETCH_FILMS_FAIL'

export const SAVE_FILM = 'SAVE_FILM'
export const SAVE_FILM_SUCCESS = 'SAVE_FILM_SUCCESS'
export const SAVE_FILM_FAIL = 'SAVE_FILM_FAIL'

export const UPDATE_FILM = 'UPDATE_FILM'
export const UPDATE_FILM_SUCCESS = 'UPDATE_FILM_SUCCESS'
export const UPDATE_FILM_FAIL = 'UPDATE_FILM_FAIL'
