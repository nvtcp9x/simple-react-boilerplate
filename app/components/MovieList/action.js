import { FETCH_FILMS, FETCH_FILMS_SUCCESS, FETCH_FILMS_FAIL, SAVE_FILM, SAVE_FILM_SUCCESS, SAVE_FILM_FAIL, UPDATE_FILM, UPDATE_FILM_SUCCESS, UPDATE_FILM_FAIL } from './constants'

export function fetchFilms() {
  return {
    type: FETCH_FILMS,
  }
}

export function fetchFilmsSuccess(data) {
  return {
    type: FETCH_FILMS_SUCCESS,
    data,
  }
}

export function fetchFilmsFail(error) {
  return {
    type: FETCH_FILMS_FAIL,
    error,
  }
}

export function saveFilm(data) {
  return {
    type: SAVE_FILM,
    data,
  }
}

export function saveFilmSuccess(data) {
  return {
    type: SAVE_FILM_SUCCESS,
    data,
  }
}

export function saveFilmFail(error) {
  return {
    type: SAVE_FILM_FAIL,
    error,
  }
}

export function updateFilm(id, data) {
  return {
    type: UPDATE_FILM,
    id,
    data,
  }
}

export function updateFilmSuccess(data) {
  console.log(data)
  return {
    type: UPDATE_FILM_SUCCESS,
    data,
  }
}

export function updateFilmFail(error) {
  return {
    type: UPDATE_FILM_FAIL,
    error,
  }
}
