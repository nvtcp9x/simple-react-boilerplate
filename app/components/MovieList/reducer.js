import { FETCH_FILMS, FETCH_FILMS_SUCCESS, FETCH_FILMS_FAIL, SAVE_FILM, SAVE_FILM_SUCCESS, SAVE_FILM_FAIL, UPDATE_FILM, UPDATE_FILM_SUCCESS, UPDATE_FILM_FAIL } from './constants'

const initialState = {
  isFetching: false,
  items: [],
  error: {},
};

export function filmReducer(state = initialState, action) {
  switch (action.type) {
    case FETCH_FILMS:
    case SAVE_FILM:
    case UPDATE_FILM:
      return {
        ...state,
        isFetching: true,
      }
    case FETCH_FILMS_SUCCESS:
    case SAVE_FILM_SUCCESS:
    case UPDATE_FILM_SUCCESS:
      return {
        ...state,
        isFetching: false,
        items: action.data,
      }
    case FETCH_FILMS_FAIL:
    case SAVE_FILM_FAIL:
    case UPDATE_FILM_FAIL:
      return {
        ...state,
        isFetching: false,
        error: action.error,
      }
    default:
      return state
  }
}
