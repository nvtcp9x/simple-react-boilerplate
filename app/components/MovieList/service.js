import axios from 'axios'
import { URL_FILMS } from './constants'

export function getFilmsApi() {
  return axios.get(URL_FILMS)
}
