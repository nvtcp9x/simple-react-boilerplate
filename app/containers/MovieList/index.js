// @flow

import * as React from 'react';
import { Link } from 'react-router-dom'
import { Col } from 'reactstrap'
import styled from 'styled-components'
import { connect } from 'react-redux'
import { URL_IMG, IMG_SIZE_LARGE } from '../../../utils/constants'
import { StyledImage, StyledRow } from '../../../utils/commonStyle'
import BookMark from '../../components/BookMark';
import { fetchFilms, saveFilm, updateFilm } from './action'
import BaseLoading from '../../components/BaseLoading'

type Props = {
  movies: Array<Object>,
  films: Object,
  fetchFilms: () => void,
  saveFilm: () => void,
  updateFilm: () => void,
  loading: boolean,
}

const Wrraper = styled.div`
  &:hover {
    .meta {
      visibility: visible
    }
  }
`

export class MovieList extends React.Component<Props> {
  componentDidMount() {
    this.props.fetchFilms();
  }

  updateBookMark(filmId) {
    const { items } = this.props.films
    const film = Array.from(items).filter(item => item.filmId === filmId)[0]
    const { id } = film;
    const data = film.bookMark ? { bookMark: false } : { bookMark: true }
    this.props.updateFilm(id, data);
  }

  saveBookMark(filmId) {
    const userId = localStorage.getItem('userId');
    const data = {
      filmId,
      userId,
      bookMark: true,
    }

    this.props.saveFilm(data)
  }

  isBookMark = (id) => {
    const { items } = this.props.films;
    const bookMarkedFilmId = Array.from(items).filter(item => item.bookMark === true).map(film => film.filmId);
    if (bookMarkedFilmId.includes(id)) {
      return true;
    }
    return false
  }

  handleBookMark = (id) => {
    const { items } = this.props.films;
    const bookMarkedFilmId = Array.from(items).map(item => item.filmId);
    if (bookMarkedFilmId.includes(id)) {
      return true;
    }
    return false
  }

  render() {
    const { movies, loading } = this.props;
    const moviesList = movies.filter(movie => movie.poster_path !== null)
      .map(movie => (
        <Col xs={6} sm={4} md={3} key={movie.id} >
          <Link to={`/movie/${movie.id}`} >
            <Wrraper>
              <StyledImage src={URL_IMG + IMG_SIZE_LARGE + movie.poster_path} className="img-fluid" />
              {localStorage.getItem('userId') ?
                <BookMark
                  checkBookMark={this.isBookMark(movie.id)}
                  handleBookMark={this.handleBookMark(movie.id)
                    ? () => this.updateBookMark(movie.id)
                    : () => this.saveBookMark(movie.id)}
                  filmId={movie.id}
                />
                : null
              }
            </Wrraper>
          </Link>
        </Col>));
    return (
      <BaseLoading loading={loading} >
        <StyledRow>
          {moviesList}
        </StyledRow>
      </BaseLoading>
    );
  }
}

export const mapStateToProps = state => ({
  films: state.filmReducer,
  loading: state.movieList.isFetching,
})

export default connect(mapStateToProps, {
  fetchFilms, saveFilm, updateFilm,
})(MovieList)
