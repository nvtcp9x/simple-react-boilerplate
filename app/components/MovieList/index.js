// @flow

import * as React from 'react';
import { Link } from 'react-router-dom'
import { Row, Col } from 'react-bootstrap'
import styled from 'styled-components'
import { connect } from 'react-redux'
import { URL_IMG, IMG_SIZE_LARGE } from '../../../utils/constants'
import { StyledImage } from '../../../utils/commonStyle'
import BookMark from '../BookMark';
import withReducer from '../../../utils/withReducer'
import { filmReducer } from './reducer'
import { fetchFilms, saveFilm, updateFilm } from './action'

type Props = {
  movies: Array<Object>,
  films: Object,
  fetchFilms: () => void,
  saveFilm: () => void,
  updateFilm: () => void,
}
const Wrraper = styled.div`
  &:hover {
    .meta {
      visibility: visible
    }
  }
`

class MovieList extends React.Component<Props> {
  state = {
    bookMark: false,
  }

  componentDidMount() {
    this.props.fetchFilms();
  }

  updateBookMark = (filmId) => {
    const userId = localStorage.getItem('userId');
    // this.setState({ bookMark: !this.state.bookMark }, () => {
    //   // if (this.state.bookMark) {
    //   //   this.props.saveFilm({ filmId, userId, bookMark: true })
    //   // }
    //   if (!this.state.bookMark) {
    //     const { id } = this.props.films
    //     console.log(this.props.films)
    //     // this.props.updateFilm(id)
    //   }
    // })
  }

  checkBookMark = (id) => {
    if (this.props.films) {
      const { items } = this.props.films;
      const bookMarkedFilmId = Array.from(items).filter(item => item.bookMark === true).map(film => film.filmId);
      if (bookMarkedFilmId.includes(id) || this.state.bookMark) {
        return true;
      }
      return false
    }
  }

  render() {
    const moviesList = this.props.movies.filter(movie => movie.poster_path !== null)
      .map(movie => (
        <Col xs={6} sm={4} md={3} key={movie.id} >
          <Link to={`/movie/${movie.id}`} >
            <Wrraper>
              <StyledImage src={URL_IMG + IMG_SIZE_LARGE + movie.poster_path} responsive />
              { localStorage.getItem('userId') ?
                <BookMark checkBookMark={this.checkBookMark(movie.id)} updateBookMark={this.updateBookMark(movie.id)} filmId={movie.id} />
                : null
              }
            </Wrraper>
          </Link>
        </Col>));
    return (
      <Row>
        {moviesList}
      </Row>
    );
  }
}

const mapStateToProps = state => ({
  films: state.filmReducer,
})

export default withReducer('filmReducer', filmReducer)(connect(mapStateToProps, { fetchFilms, saveFilm, updateFilm })(MovieList))
