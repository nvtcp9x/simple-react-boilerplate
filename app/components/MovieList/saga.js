import { takeEvery, put, call } from 'redux-saga/effects';
import { FETCH_FILMS } from './constants';
import { fetchFilmsSuccess, fetchFilmsFail } from './action';
import { getFilmsApi } from './service';

export function* fetchFilms() {
  try {
    const response = yield call(getFilmsApi);
    yield put(fetchFilmsSuccess(response.data));
  } catch (err) {
    yield put(fetchFilmsFail(err));
  }
}

export function* watchFetchFilms() {
  yield takeEvery(FETCH_FILMS, fetchFilms)
}
