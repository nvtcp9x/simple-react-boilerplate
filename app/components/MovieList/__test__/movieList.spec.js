import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import MovieList from '../index';

Enzyme.configure({ adapter: new Adapter() });

describe('<MovieList />', () => {
  const mockProps = {
    movies: [
      {
        id: 1,
        poster_path: 'jkfdjfd',
      },
    ],
  }
  const component = shallow(<MovieList {...mockProps} />)

  it('should return div tag', () => {
    expect(toJson(component)).toMatchSnapshot();
  })
})
