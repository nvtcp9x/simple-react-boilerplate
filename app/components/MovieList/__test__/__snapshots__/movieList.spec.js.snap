// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieList /> should return div tag 1`] = `
<Row
  bsClass="row"
  componentClass="div"
>
  <Col
    bsClass="col"
    componentClass="div"
    key="1"
    md={3}
    sm={4}
    xs={6}
  >
    <Link
      replace={false}
      to="/movie/1"
    >
      <styled.div>
        <Styled(Image)
          responsive={true}
          src="https://image.tmdb.org/t/p/w342/jkfdjfd"
        />
      </styled.div>
    </Link>
  </Col>
</Row>
`;
